From 49f24d231f6b9e25daac1df5a8720b4e72d66c00 Mon Sep 17 00:00:00 2001
From: David Edmundson <kde@davidedmundson.co.uk>
Date: Thu, 20 Nov 2025 17:49:34 +0000
Subject: [PATCH] inputmethod: Skip input events from fake input

On SteamOS the Steam OSK internally uses ibus and commits complete
keysms that should go to the client.

If kwin believes another grabbing InputMethod is enabled we currently
then forward the complete key symbol to the input method which then ends
up getting lost.

This patch allows input from a physical to use the configured input
method as usual, but skips input that comes from libei.
---
 src/input.cpp | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/input.cpp b/src/input.cpp
index 849ff36301..48fdb66e18 100644
--- a/src/input.cpp
+++ b/src/input.cpp
@@ -254,6 +254,13 @@ bool InputEventFilter::passToInputMethod(KeyboardKeyEvent *event)
     if (!kwinApp()->inputMethod()) {
         return false;
     }
+
+    if (event->device && event->device->metaObject()->className() == QLatin1String("KWin::EisDevice")) {
+        // do not pass synthetic events to input method, these could be from steam's OSK
+        // which have already come from a ibus
+        return false;
+    }
+
     if (auto keyboardGrab = kwinApp()->inputMethod()->keyboardGrab()) {
         const auto timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(event->timestamp);
         keyboardGrab->sendKey(waylandServer()->display()->nextSerial(), std::chrono::duration_cast<std::chrono::milliseconds>(timestamp).count(), event->nativeScanCode, event->state);
-- 
2.51.2

